### DISCUSSION TITLE
Converting MarkupExtension to use Compiled Bindings

### DISCUSSION THREAD

**Discussion by IsaacMarovitz** [AvaloniaUI/Avalonia#12411] [Q&A]
Hi y'all. I have the following markup extension:

```csharp
internal class LocaleExtension : MarkupExtension
{
    public LocaleExtension(LocaleKeys key)
    {
        Key = key;
    }

    public LocaleKeys Key { get; }

    public override object ProvideValue(IServiceProvider serviceProvider)
    {
        LocaleKeys keyToUse = Key;

        ReflectionBindingExtension binding = new($"[{keyToUse}]")
        {
            Mode = BindingMode.OneWay,
            Source = LocaleManager.Instance,
        };

        return binding.ProvideValue(serviceProvider);
    }
}
```

I'm looking to convert it to a Compiled Binding. My initial attempt was to return the value, without any use of the BindingExtensions, from LocaleKeys, but that did not update when the locale was changed at runtime. Any suggestions?

**Comment by timunie** [comment]
just return a `Binding`

```cs
public override object ProvideValue(IServiceProvider serviceProvider)
{
    LocaleKeys keyToUse = Key;

    Binding binding = new($"[{keyToUse}]")
    {
        Mode = BindingMode.OneWay,
        Source = LocaleManager.Instance,
    };

    return binding);
}
```

**Reply by IsaacMarovitz** [reply]
That still uses a Reflection Binding under the hood: `Using member 'Avalonia.Data.Binding.Binding(String, BindingMode)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. BindingExpression and ReflectionBinding heavily use reflection. Consider using CompiledBindings instead.`

**Comment by workgroupengineering** [comment]
You can use like this:

```c#
var path = new CompiledBindingPathBuilder().
  ...
  .Build();

return new CompiledBindingExtension(path)
{
   Mode = BindingMode.OneWay,
};
```

**Reply by IsaacMarovitz** [reply]
The part I don't understand is how to use CompiledBindingPathBuilder. Are there any examples you can link?

**Reply by IsaacMarovitz** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/42140194/d850ef68-5f19-423e-88fb-cbbeb0a8498d)

This was my attempt, it did not work.

**Reply by IsaacMarovitz** [reply]
In the debugger, it seems it is passing the ViewModel as the obj not the LocaleKeys

**Reply by IsaacMarovitz** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/42140194/d5a5fd72-1a26-478c-8c65-3146926acd44)

This is how the extension is used

**Reply by jp2masa** [reply]
This is a binding, so `obj` is the data context. If you need the `LocaleKeys`, you have the `Key` property in your markup extension class, you can just use it.

**Reply by IsaacMarovitz** [reply]
That's what I did originally, but that will not update if the Locale is changed for example.

**Reply by jp2masa** [reply]
Then you need to bind to the Locale property. I would need to see how Locale is defined to provide some example code.

**Reply by timunie** [reply]
@IsaacMarovitz try to prepare a very minimal sample of what you have so far and upload / link it here. Probably someone will find a solution hopefully.

**Reply by Symbai** [reply]
@workgroupengineering Thank you. Came here because I had the same issue with localization service binding and I can confirm that this is the solution.

**Reply by maxkatz6** [reply]
New [Binding.Create](https://github.com/AvaloniaUI/Avalonia/pull/20443) API was recently merged into master. To be part of v12, should be available in current nightly builds.
